fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,647 @@
+                                              ffg_support_for_checkpoint: int,
+                                              min_ffg_support_slash_th: int) -> int:
+    assert min_ffg_support_slash_th <= ffg_support_for_checkpoint
+    """
+    Return the max possible ffg_confirmation_score such that:
+    2 / 3 * total_active_balance <= \
+        ffg_support_for_checkpoint - \
+        min(min_ffg_support_slash_th, total_active_balance * ffg_confirmation_score / 100)
+    """
+
+    """
+    Case 1: min_ffg_support_slash_th <= total_active_balance * ffg_confirmation_score / 100
+
+    2 / 3 * total_active_balance <= \
+        ffg_support_for_checkpoint - min_ffg_support_slash_th
+    """
+    if 2 * int(total_active_balance) <= (ffg_support_for_checkpoint - min_ffg_support_slash_th) * 3:
+        return 100 // 3
+
+    """
+    Case 2: min_ffg_support_slash_th > total_active_balance * ffg_confirmation_score / 100
+
+    2 / 3 * total_active_balance <= \
+        ffg_support_for_checkpoint - total_active_balance * ffg_confirmation_score / 100
+    """
+    ffg_confirmation_score = (
+        (300 * (ffg_support_for_checkpoint) - 200 * int(total_active_balance)) //
+        (3 * (total_active_balance))
+    )
+    return max(ffg_confirmation_score, -1)
+```
+
+#### `get_ffg_confirmation_score`
+
+```python
+def get_ffg_confirmation_score(store: Store, 
+                               confirmation_slashing_threshold: int,
+                               block_root: Root) -> int:
+
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    total_active_balance = int(get_total_active_balance_for_block_root(store, block_root))
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, block_root))
+
+    min_ffg_support_slash_th = min(ffg_support_for_checkpoint, confirmation_slashing_threshold)
+
+    if block_epoch == current_epoch:
+        remaining_ffg_weight = int(get_remaining_weight_in_current_epoch(store, block_root))
+        return get_ffg_confirmation_score_current_epoch(
+            total_active_balance,
+            ffg_support_for_checkpoint,
+            min_ffg_support_slash_th,
+            remaining_ffg_weight
+        )
+    else:
+        return get_ffg_confirmation_score_previous_epoch(
+            total_active_balance,
+            ffg_support_for_checkpoint,
+            min_ffg_support_slash_th
+        )
+```
+
+### `get_confirmation_score`
+
+```python
+def get_confirmation_score(store: Store,
+                           confirmation_slashing_threshold: int,
+                           block_root: Root) -> int:
+    """
+    Return -1 in the case that `block_root` cannot be confirmed even by assuming no adversary weight,
```

Comment:
Why don't it return `0` as this would be a size of an adversary a block can tolerate in the case when it is unconfirmed?

Reply:
This is the logic that I have followed:
- 0 would mean that the block is confirmed, assuming no adversarial stake
- -1 indicates that the block cannot be confirmed even if one assumes adversarial stake